app/models/thredded/user_topic_follow.rb

```diff
@@ -13,8 +13,19 @@ class UserTopicFollow < ActiveRecord::Base
     alias_attribute :postable_id, :topic_id
     alias_attribute :postable, :topic
 
-    def self.create_unless_exists!(user_id, topic_id, reason = :manual)
-      create_with(reason: reason).find_or_create_by!(user_id: user_id, topic_id: topic_id)
+    # Creates a follow or finds the existing one.
+    #
+    # This method is safe to call concurrently from different processes. Lookup and creation happen in a transaction.
+    # If an ActiveRecord::RecordNotUnique error is raised, the find is retried.
+    #
+    # @return [Thredded::UserTopicFollow]
+    def self.create_unless_exists(user_id, topic_id, reason = :manual)
+      transaction(requires_new: true) do
+        create_with(reason: reason).find_or_create_by(user_id: user_id, topic_id: topic_id)
+      end
+    rescue ActiveRecord::RecordNotUnique
+      # The record has been created from another connection, retry to find it.
+      retry
```

Comment:
> And if I understand correctly, the new transaction is to ensure that the isolation happens immediately 

Yes

> But if so, then there is no need for a retry -- the retry will only successfully find the record.

The record could be deleted in the meantime, in which case the retry will create one. Also, it's nice to have a consistent return type, i.e. this method is guaranteed to return a UserTopicFollow.